src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -112,9 +134,27 @@ object GroupReadsByUmi {
       override def equals(other: scala.Any): Boolean = other.isInstanceOf[Node] && this.umi == other.asInstanceOf[Node].umi
     }
 
-    override def assign(rawUmis: Seq[String]): Map[String, String] = {
+    /** Returns the count of each raw UMI that was observed. */
+    protected def count(umis: Seq[Umi]) : Iterator[(Umi,Long)] = SimpleCounter(umis).iterator
+
+    /** Returns the differences between a pair of UMIs. */
+    protected def differences(lhs: Umi, rhs: Umi): Int = Sequences.countMismatches(lhs, rhs)
+
+    /** Assigns IDs to each UMI based on the root to which is it mapped. */
+    protected def assignIds(roots: Seq[Node]) : Map[Umi, MoleculeId] = {
+      val mappings = mutable.Buffer[(Umi,MoleculeId)]()
+      roots.foreach(root => {
+        val id = nextId
+        mappings.append((root.umi, id))
+        root.descendants.foreach(child => mappings.append((child.umi, id)))
```

Comment:
While more scala-ish, I think what's there is ok and avoids a bunch of temporary map creation.